### ISSUE TITLE
Different coverage in free function vs associated methods

### ISSUE THREAD

**Issue by pustaczek** [xd009642/tarpaulin#238]
Tarpaulin 0.8.4 reports different coverage depending on whether the code is inside a free function or an associated method. Specifically, inside a free function a [`str::lines`](https://doc.rust-lang.org/std/primitive.str.html#method.lines) call is not included in the output at all, while in an associated method it is reported as missed. The following is a minimal example, originally found [here](https://coveralls.io/builds/24262895/source?filename=src/paste/piece_parse.rs#L9):
```rust
struct S;

impl S {
    fn f(a: &str) {
        let _ = a
            .lines()
            .collect::<Vec<_>>();
    }
}

#[test]
fn t() {
    S::f("a\nb\n");
}
```
```rust
fn f(a: &str) {
    let _ = a
        .lines()
        .collect::<Vec<_>>();
}

#[test]
fn t() {
	f("a\nb\n");
}
```
![(Tarpaulin HTML output screenshot)](https://imgur.com/uBNqiHt.png)

Ubuntu 19.04, Linux 5.0.0-20-generic. Issue also occurs on Travis CI.
rustc 1.37.0-nightly (5f9c0448d 2019-06-25)

**Comment by jstasiak** [comment]
I think I've also bumped into this (both `zero` and `unitx` methods are used):

<img width="1084" alt="Screenshot 2019-07-09 at 14 57 04" src="https://user-images.githubusercontent.com/36209/60889701-eb00d480-a259-11e9-81d8-9c47a04271b7.png">

https://coveralls.io/builds/24447374/source?filename=src/lib.rs#L27

The full source code: https://github.com/jstasiak/raytracer/blob/818786af2ec9f7622abdb0c4ca1892a90d87a797/src/lib.rs#L19

**Comment by jstasiak** [comment]
Not sure if related, but in the following code one line is missing from the coverage and I'm only using free functions here:

```rust
pub fn fun1() -> i32 {
	1
}

pub fn fun2() -> i32 {
	1 + 1
}

#[test]
fn it_works() {
	let _ = fun1();
	let _ = fun2();
}
```

<img width="328" alt="Screenshot 2019-07-09 at 19 16 03" src="https://user-images.githubusercontent.com/36209/60909312-3e851980-a27e-11e9-89b2-ada66cf7df2c.png">

**Comment by xd009642** [comment]
If I had to guess I'd say `fun1` is likely inlined by the compiler so doesn't show up as an instrumentable line whereas as `fun2` contains an expression it's not inlined with the compiler flags as they are

**Comment by jstasiak** [comment]
I don't think it's gonna be inlined per se (as the default build is a debug one, I believe? I'm not adding any extra options to `cargo tarpaulin` invocation) but other than inlining I have zero idea what's going on here. I've found something interesting in `cargo-asm` output of those two functions though, may be of interest to you: https://github.com/gnzlbg/cargo-asm/issues/129.

**Comment by pustaczek** [comment]
The inlining does seem to be the cause as confirmed by adding `#[inline(never)]`, but the output is still interesting. Both `.lines()` calls in my example are still marked as missed, and in the other example, the line returning the `1` literal is marked as missed while `1+1` is not.

In the `--debug` tarpaulin output, these three lines are marked as "potentially uncoverable". In tarpaulin's sources I can see some logic responsible for ignoring some false positives, which probably means it just does not handle these cases. I am not familiar with the DWARF format though, so I don't know why it fails on these lines specifically and how to fix it.

![(screenshot of tarpaulin report)](https://i.imgur.com/ahFeIFJ.png)

**Comment by xd009642** [comment]
Closed in favour of #351
